controllers/etcdcluster_controller.go

```diff
@@ -80,17 +87,131 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 		}
 		service = headlessServiceForCluster(cluster)
 		if err := r.Create(ctx, service); err != nil {
-			log.Error(err, "unable to create Service", "namespace", service.Namespace, "cluster", cluster.Name)
+			log.Error(err, "unable to create Service", "service", service.Name)
+			return ctrl.Result{}, err
+		}
+		log.V(1).Info("Created Service", "service", service.Name)
+		return ctrl.Result{}, nil
+	}
+	log.V(2).Info("Service exists", "service", service.Name)
+
+	peers := &etcdv1alpha1.EtcdPeerList{}
+	if err := r.List(ctx, peers, client.MatchingFields{clusterNameSpecField: cluster.Name}); err != nil {
+		log.Error(err, "unable to list peers")
+		return ctrl.Result{}, err
+	}
+
+	// Assume we can't communicate with the cluster
+	log.V(1).Info("Unable to communicate with etcd cluster")
+	if int32(len(peers.Items)) < *cluster.Spec.Replicas {
```

Comment:
This crashes the operator as follows:

```
2019-10-18T11:16:59.647Z	DEBUG	controllers.EtcdCluster	Unable to communicate with etcd cluster	{"cluster": "default/my-cluster"}
E1018 11:16:59.648187       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:76
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:65
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:199
/usr/local/go/src/runtime/signal_unix.go:394
/workspace/controllers/etcdcluster_controller.go:106
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:216
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:192
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:171
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x10dd7e8]

goroutine 302 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1247dc0, 0x1e85a50)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/improbable-eng/etcd-cluster-operator/controllers.(*EtcdClusterReconciler).Reconcile(0xc00059b380, 0xc0005248c9, 0x7, 0xc0005248b0, 0xa, 0x40e700, 0x0, 0x0, 0x0)
	/workspace/controllers/etcdcluster_controller.go:106 +0xa68

```

When I tested it with @adamhosier 's e2e test suite.

We should do defaulting of these values in three places: 
1. The CRD definition
2. A webhook 
3. And in case the webhook isn't running, we should run the same defaulting functions right after we've got the cluster, at the start of each Reconciliation.

Reply:
I've added a check for null pointer here. I don't default, just don't crash if you don't set it.

Reply:
Ok. I created https://github.com/improbable-eng/etcd-cluster-operator/issues/39 as a followup.